src/Microsoft.SqlTools.ServiceLayer/Metadata/SmoScripterHelpers.cs

```diff
@@ -0,0 +1,206 @@
+﻿//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using System.Collections.Generic;
+using System.Data.Common;
+using System.Threading.Tasks;
+using Microsoft.Data.SqlClient;
+using Microsoft.SqlServer.Management.Common;
+using Microsoft.SqlServer.Management.Smo;
+using Microsoft.SqlTools.SqlCore.Connection;
+using Microsoft.SqlTools.ServiceLayer.Connection.ReliableConnection;
+using System;
+using Microsoft.SqlTools.Utility;
+
+namespace Microsoft.SqlTools.ServiceLayer.Metadata
+{
+    internal static class SmoScripterHelpers
+    {
+        public static async Task<IEnumerable<string>?> GenerateAllServerTableScripts(DbConnection connection)
+        {
+            var serverConnection = SmoScripterHelpers.GetServerConnection(connection);
+            if (serverConnection == null)
+            {
+                return null;
+            }
+
+            Server server = new Server(serverConnection);
+            var scripts = await SmoScripterHelpers.GenerateTableScripts(server);
+
+            return scripts;
+        }
+
+        private static ServerConnection? GetServerConnection(DbConnection connection)
+        {
+            // Get a connection to the database for SMO purposes
+            var sqlConnection = connection as SqlConnection ?? SmoScripterHelpers.TryFindingReliableSqlConnection(connection as ReliableSqlConnection);
+            if (sqlConnection == null)
+            {
+                return null;
+            }
+
+            var serverConnection = SmoScripterHelpers.ConnectToServerWithSmo(sqlConnection);
+            return serverConnection;
+        }
+
+        private static SqlConnection? TryFindingReliableSqlConnection(ReliableSqlConnection reliableSqlConnection)
+        {
+            // It's not actually a SqlConnection, so let's try a reliable SQL connection
+            if (reliableSqlConnection == null)
+            {
+                // If we don't have connection we can use with SMO, just give up on using SMO
+                return null;
+            }
+
+            // We have a reliable connection, use the underlying connection
+            return reliableSqlConnection.GetUnderlyingConnection();
+        }
+
+        private static ServerConnection ConnectToServerWithSmo(SqlConnection connection)
```

Comment:
I've seen stuff like this around before, if there isn't a common helper method already to do this then it seems useful to make one.

Reply:
This is the only other spot I could find that does something similar, but it's getting the server name and not the server itself.

https://github.com/microsoft/sqltoolsservice/blob/16c7e323567023b08200ab35b2bfb2a5042385e0/src/Microsoft.SqlTools.ServiceLayer/Scripting/SmoScriptingOperation.cs#L90C16-L98C21